Hybrid analysis bug and commands

reutshal opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Status
In Progress

## Related Issues
fixes: https://github.com/demisto/etc/issues/17018
fixes: https://github.com/demisto/etc/issues/17154

## Description
1. Bug fixed - Calculation for DbotScore added.
3. Customer requested commands was implemented.

branch | PR
------ | ------

## Does it break backward compatibility?
   - No

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [X ] Code Review

## Additional changes
1. Another DBot score calculation was added (quick_scan_result_to_entry function).
2. Fixed another bug founded- didn't add malicious entry to context.

guykeller commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,6 +363,186 @@ script:
+        if (malicious_counter >= threshold)
+            Score = 3;
+
+        var Dbot_context = {
```
There should also be a "Type" field here

guykeller commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,6 +363,186 @@ script:
+            'Vendor': "Hybrid Analysis",
+            'Score': Score,
+            }
+        var whitelist = {
```
Could you please explain what this variable means?

guykeller commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,6 +363,186 @@ script:
+            }
+        var whitelist = {
+            'ID': response.whitelist[0].id,
+            'Value' : response.whitelist[0].value
```
please split this to several lines for readability and clarity.
For example:
whitelist = response.whitelist
first_whitelisted_file = whitelist[0]
then the ID and Value.
However, this note might change after I understand the variable's meaning better

guykeller commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,6 +363,186 @@ script:
+        if (Score == 3){
+            res.Malicious = {
+                Vendor: 'Hybrid Analysis',
+                Description: 'DBotScore is 3' // check this!
```
This does not seem like the correct description, have you checked it?

reutshal replied:
This is the definition :
        "Malicious": {
             "Vendor": "STRING, Vendor reporting malicious",
             "Description": "STRING, Description about why File was determined malicious"}
I just didn't know what to write as the reason (this is the calculation by design) so wanted to leave it to Alex. Ideas?

guykeller commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,6 +363,186 @@ script:
+            }
+            // if (Score == 3)
+            //     context.URL.Malicious = res;
+        }
```
I think that a good coding practice is to try to keep functions short.
For example, this function is about 80 lines long - making it very difficult to see it's purpose in a quick look.
If we export smaller purposes to other functions we could make this function shorter and clearer

guykeller commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,6 +363,186 @@ script:
+    }
+
+
+    function quick_scan_to_entry(response, indicator, url) {
```
might be clearer if we rename indicator to indicatorType.
Also, an explaining comment could come in handy here, since currently the function is complicated (since the product's logic is complicated inherently)

reutshal replied:
Done

guykeller commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -373,211 +562,430 @@ script:
-          description: Country (3 digit ISO) e.g. swe
-        - name: verdict
-          description: 'Verdict e.g. 1 (available: 1 ''whitelisted’, 2 ''no verdict’,
+  - name: hybrid-analysis-scan
```
is there a reason all the lines below this one (including this one) used to be tabbed out and now they aren't?

guykeller replied:
@reutshal had a chance to look at this?

ronykoz replied:
@reutshal Upload the yml yo demisto to make sure we are good here.

reutshal replied:
Looking good

guykeller commented:
General notes:
1. Try to keep functions short and concise - generally speaking, every function should be in charge of one thing. That one thing could be a big thing, in which case that function will "manage" the work of a bunch of smaller functions with indicative names
2. Please change naming to JS conventions and not Python conventions

And one tip - when changing something in the CR, pay close attention and make sure you've changed it everywhere. I've made this mistake so many times after CRs and it's frustrating to fix afterwards

reutshal commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -38,6 +37,8 @@ script:
     var apiKey;
     var integrationContext = getIntegrationContext();
     var licenseID = getLicenseID();
+    var Dbot = [1,2,3,0]; //  threat_level : (0=No Threat, 1=Suspicious, 2=Malicious, 3=Unknown)
```
Implemented option 2

reutshal commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -76,13 +77,19 @@ script:
 
     // return a function that maps object keys by mapper (or capitlize keys if key is not exists in mapper)
-    function mapObject(mapper) {
+    function mapObject(mapper, identifier) {
```
Done

reutshal commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -94,7 +101,7 @@ script:
             Contents: rawResponse,
             ReadableContentsFormat: formats.markdown,
-            HumanReadable: tableToMarkdown(name, table, headers, undefined, headerTransform=underscoreToCamelCase, removeNull=true),
+            HumanReadable: tableToMarkdown(name, table, headers, undefined, headerTransform=undefined, removeNull=true),
```
Without the change the human readable appeared with underscore: _EnvironmentID . 
Wanted to fix this

reutshal commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -147,30 +155,33 @@ script:
            isurlanalysis: 'url analysis',
            analysis_start_time: 'analysis start time',
-           total_signatures: 'total signatures'
+           total_signatures: 'total signatures',
```
Not needed. deleted

ronykoz commented:
@reutshal for the future - it is better to split it into two PRs, this way the code review process would have been faster and maybe some of the changes would have entered the upcoming release

reutshal commented:
@ronykoz Totally right, Keller and I talked about this.

guykeller commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,230 +431,631 @@ script:
+        var dbotScore = calculateDbotScore(unknownCounter, maliciousCounter, minMaliciousScanners, response.scanners.lenght);
+
+        var context = createQuickScanContext (response, scannersContext, dbotScore, whitelist);
+        var hr = createQuickScanHumanReadable (response, scannersContext, whitelist);
```
this name could be more clear, maybe humanReadable?

guykeller commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,230 +431,631 @@ script:
+
+        var context = createContext(response, hybrid_context)
+
+        // adding { ID : url } to the integration context for future use
```
Did not understand this comment

ronykoz replied:
@reutshal  Why?

guykeller commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,230 +431,631 @@ script:
+        var context = createContext(response, hybrid_context)
+
+        // adding { ID : url } to the integration context for future use
+        var add_to_context = {};
```
What is the meaning of this variable? it should be something like newIntegrationContext, so the name of the variable reflects it's role.
Moreover, this is Python naming. Should be addToContext

ronykoz commented:
@reutshal lets get finish this PR, it is open for too long

reutshal commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,230 +431,631 @@ script:
   runonce: false
 tests:
-- "Detonate File - HybridAnalysis - Test"
+  - HybridAnalysis-Test
```
Thought it was wrong and that this entry should contain only the test playbook.
Will add this back

reutshal commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,230 +431,631 @@ script:
-            var entry = searchQuery('url:google');
-            if (entry && entry.Type === entryTypes.note) {
+            var res = quickScanUrl('all', 'www.google.com', '/api/v2/quick-scan/url');
+            if (res != null && res.sha256 == 'd2edef8e43054be586d17ddcc761e7a1f4a6946c39e653d7e095a826ef34b6a1') {
```
changed it to 'submission_type' check

reutshal commented:
@kirbles19 could you go over the following files:
https://github.com/demisto/content/blob/hybrid-analysis-bug-commands/Integrations/integration-HybridAnalysis.yml

https://github.com/demisto/content/blob/hybrid-analysis-bug-commands/Integrations/integration-HybridAnalysis_CHANGELOG.md

specifically over the description of 'malicious_threat_levels' in hybrid-analysis-scan,
'min_malicious_scanners' in hybrid-analysis-quick-scan-url-results.

thanks

kirbles19 commented:
@reutshal All done reviewing .yml and .md. When you approve the changes, please add the `docs-approved` label to the PR.

ronykoz commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,230 +431,639 @@ script:
+      required: true
+      description: The MD5, SHA1, or SHA256 hash of the file.
+    - name: malicious_threat_levels
+      default: true
```
why is it default?

reutshal replied:
removed the default

ronykoz commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,230 +431,639 @@ script:
+      description: The MD5, SHA1, or SHA256 hash of the file.
+    - name: malicious_threat_levels
+      default: true
+      description: Threat level values, which will determine if the file is malicious.
```
@reutshal is it a good enough explanation?

ronykoz replied:
Better to talk with Alex about this one

reutshal replied:
Changing this to :
A CSV list of threat level values. Files that have a threat level specified in the list will be considered malicious.

ronykoz commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,230 +431,639 @@ script:
+  - name: hybrid-analysis-quick-scan-url
+    arguments:
+    - name: scan_type
+      default: true
```
remove the default here

ronykoz commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,230 +431,639 @@ script:
+      required: true
+      description: The ID of the scanned URL.
+    - name: min_malicious_scanners
+      default: true
```
remove the default it it not needed here

ronykoz commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,230 +431,639 @@ script:
+            Contents: result,
+            ReadableContentsFormat: formats.markdown,
+            HumanReadable: '### Scanner:\n' + tableToMarkdown(null, scannersContext, ['Name', 'Available', 'Description'], undefined, headerTransform=undefined, removeNull=true),
+            EntryContext: {'HybridAnalysis.Scanner(val.name && val.name == obj.name) :' : scannersContext}
```
```suggestion
            EntryContext: {'HybridAnalysis.Scanner(val.Name && val.Name == obj.Name) :' : scannersContext}
```

ronykoz commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,230 +431,639 @@ script:
+        //Sandbox Submission commands
+        case 'hybrid-analysis-submit-url':
+            var response = submitUrl(args.url, args.environmentID, '/api/v2/submit/url' );
+            return submitFileMsg(response);
```
is this function really suits your needs?

reutshal commented:
@kirbles19 I changed the 'ID' output into 'ScanID' in 'quick-scan-url' , 'quick-scan-url-results' commands.
Also updated the 'malicious' entry description in line 495

mgioia-demisto commented:
following..

kirbles19 commented:
> quick-scan-url

In line 495, make URL uppercase please.

kirbles19 commented:
@reutshal Outputs looked good. In the message in line 495, please change `url` to uppercase `URL`.

ronykoz commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,230 +431,643 @@ script:
+        addToContext[url] = response.submission_type;
+        setIntegrationContext(addToContext);
+
+        return createTableEntry('Scan information:', response,
```
why don't you use the hybrid_context ?
I think we should show the URL as well

ronykoz commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -354,230 +431,643 @@ script:
+    //create table entry from response after 'quick-scan-url-results' command
+    function quickScanResultToEntry(response, minMaliciousScanners) {
+
+        var scannersContext = createScannersContext (response)['scannersContext'];
```
you are running the same function 4 times with the same param, better to run it once and extract the info later

guykeller commented on Integrations/integration-HybridAnalysis.yml:
```diff
@@ -75,26 +74,86 @@ script:
     }
 
+
+    function fixContextOutputs(res) {
```
Thanks for the explanation, definitely clarified a few things.
However, I still think that "res" here is naming we can improve. What do you think?

guykeller replied:
I also think the naming of this function is tricky, as we're not fixing current con text outputs - we're adding new and better outputs.
Therefore I believe this function should be renamed

ronykoz commented on Playbooks/playbook-Hybrid-analysis_quick-scan.yml:
```diff
@@ -0,0 +1,204 @@
+id: Hybrid-analysis quick-scan
+version: -1
+fromversion: 5.0.0
```
Let's verify this, i think it can be 4.5 or even 4.1

reutshal replied:
changed to 4.1
